pkg/estimator/server.go

```diff
@@ -0,0 +1,302 @@
+/*
+Copyright 2022 The Clusternet Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package estimator
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/informers"
+	informerv1 "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	listerv1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/controller-manager/pkg/clientbuilder"
+	"k8s.io/klog/v2"
+
+	appsapi "github.com/clusternet/clusternet/pkg/apis/apps/v1alpha1"
+	clusterapi "github.com/clusternet/clusternet/pkg/apis/clusters/v1beta1"
+	schedulerapi "github.com/clusternet/clusternet/pkg/apis/scheduler"
+	framework "github.com/clusternet/clusternet/pkg/estimator/framework/interfaces"
+	"github.com/clusternet/clusternet/pkg/estimator/framework/plugins"
+	frameworkruntime "github.com/clusternet/clusternet/pkg/estimator/framework/runtime"
+	"github.com/clusternet/clusternet/pkg/known"
+	"github.com/clusternet/clusternet/pkg/scheduler/parallelize"
+	"github.com/clusternet/clusternet/pkg/utils"
+)
+
+const (
+	nodeNameKeyIndex = "spec.nodeName"
+)
+
+// GetPodsAssignedToNodeFunc is a function which accept a node name input
+// and returns the pods that assigned to the node.
+type GetPodsAssignedToNodeFunc func(string) ([]*corev1.Pod, error)
+
+type EstimatorServer struct {
+	ctx context.Context
+
+	// controllerOptions for leader election and client connection
+	controllerOptions *utils.ControllerOptions
+
+	kubeClient kubernetes.Interface
+
+	informerFactory informers.SharedInformerFactory
+	nodeInformer    informerv1.NodeInformer
+	podInformer     informerv1.PodInformer
+	nodeLister      listerv1.NodeLister
+	podLister       listerv1.PodLister
+
+	getPodFunc GetPodsAssignedToNodeFunc
+
+	// default in-tree registry
+	registry  frameworkruntime.Registry
+	framework framework.Framework
+}
+
+func NewEstimatorServer(
+	ctx context.Context,
+	opts *utils.ControllerOptions,
+) (*EstimatorServer, error) {
+	clientConfig, err := utils.LoadsKubeConfig(&opts.ClientConnection)
```

Comment:
Please loads kubeconfig outside of this function because the upper function may not come from cmd.

Reply:
Done